### ISSUE TITLE
UP007 rule is not catching type aliases defined without the `TypeAlias` annotation

### ISSUE THREAD

**Issue by ericbn** [charliermarsh/ruff#4108]
Hi! 👋 

The UP007 rule is not catching the scenarios where a type alias is defined without the `TypeAlias` annotation, which is likely to happen when upgrading from py39 to py310.
 
#### test.py
Minimal code snipped that reproduces the bug:
```
from typing import Literal, Optional, TypeAlias, Union

YesNo = Literal["yes", "no"]
OptionalYesNo = Optional[YesNo]
OptionalBlankYesNo = Union[OptionalYesNo, Literal[""]]
OptionalYesNoAlias: TypeAlias = Optional[YesNo]
OptionalBlankYesNoAlias: TypeAlias = Union[OptionalYesNo, Literal[""]]
```

#### Ruff output

The command I've invoked: `ruff --isolated --select UP --target-version py310 test.py`
```
test.py:6:33: UP007 [*] Use `X | Y` for type annotations
test.py:7:38: UP007 [*] Use `X | Y` for type annotations
Found 2 errors.
[*] 2 potentially fixable with the --fix option.
```

I was expecting UP007 to have reported lines 4 and 5 too. When using `--fix` I was expecting the fixed code to be:
```
from typing import Literal, TypeAlias

YesNo = Literal["yes", "no"]
OptionalYesNo = YesNo | None
OptionalBlankYesNo = OptionalYesNo | Literal[""]
OptionalYesNoAlias: TypeAlias = YesNo | None
OptionalBlankYesNoAlias: TypeAlias = OptionalYesNo | Literal[""]
```

#### ruff --version
```
ruff 0.0.262
```

**Comment by charliermarsh** [comment]
Hmm yeah, I was confused too here, but looking back at the code, I think this is by design... We have to be somewhat careful about transforming runtime unions as those transformations can lead to subtle bugs. This was one example: https://github.com/charliermarsh/ruff/issues/3215. It also came up here: https://github.com/charliermarsh/ruff/issues/2981.

I need to see whether this is safe to do in more contexts. We could also consider flagging those usages even if we don't autofix them.

(We explicitly special-case the RHS of an assignment annotated with `TypeAlias`.)

**Comment by ericbn** [comment]
Fair enough. I cannot even start to wonder all the possible edge cases where transforming type aliases can lead to bugs after looking at the issues you've mentioned.

Sorry, I should have researched more on past issues before opening this one.

Unless you still want to consider researching on possible safe scenarios, this issue can be closed.

**Comment by ericbn** [comment]
FYI I didn't check pyupgrade's behavior before opening this issue and doing `pyupgrade --py310-plus test.py` actually does not affect the file, not even lines 6 and 7.

**Comment by charliermarsh** [comment]
Haha yeah, I think we take slightly more liberties when `TypeAlias` is added explicitly. I'm going to close for now, though I would like to revisit this at some point.
